hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/InsertIntoHoodieTableCommand.scala

```
@@ -66,100 +79,139 @@ object InsertIntoHoodieTableCommand extends Logging with ProvidesHoodieConfig {
+    val cleanedQuery = stripMetaFields(query)
+    // To validate and align properly output of the query, we simply filter out partition columns with already
+    // provided static values from the table's schema
+    //
+    // NOTE: This is a crucial step: since coercion might rely on either of a) name-based or b) positional-based
+    //       matching it's important to strip out partition columns, having static values provided in the partition spec,
+    //       since such columns wouldn't be otherwise specified w/in the query itself and therefore couldn't be matched
+    //       positionally for example
+    val expectedQueryColumns = catalogTable.tableSchemaWithoutMetaFields.filterNot(f => staticPartitionValues.contains(f.name))
+    val coercedQueryOutput = coerceQueryOutputColumns(StructType(expectedQueryColumns), cleanedQuery, catalogTable, conf)
+
+    val staticPartitionValuesExprs = createStaticPartitionValuesExpressions(staticPartitionValues, targetPartitionSchema, conf)
+
+    Project(coercedQueryOutput.output ++ staticPartitionValuesExprs, coercedQueryOutput)
+  }
+
+  private def coerceQueryOutputColumns(expectedSchema: StructType,
+                                       query: LogicalPlan,
+                                       catalogTable: HoodieCatalogTable,
+                                       conf: SQLConf): LogicalPlan = {
+    val planUtils = sparkAdapter.getCatalystPlanUtils
+    try {
+      planUtils.resolveOutputColumns(catalogTable.catalogTableName, expectedSchema.toAttributes, query, byName = true, conf)
+    } catch {
+      // NOTE: In case matching by name didn't match the query output, we will attempt positional matching
+      case ae: AnalysisException if ae.getMessage().startsWith("Cannot write incompatible data to table") =>
+        planUtils.resolveOutputColumns(catalogTable.catalogTableName, expectedSchema.toAttributes, query, byName = false, conf)
+    }
```

do we throw here if some other exception other than `AnalysisException` is thrown ? looks like we swallow silently. can we fix that?

---

Yes. It's a partial function that in case of no match will just let exception fly out, it doesn't swallow anything but AE